client/app/pages/queries/hooks/useQueryExecute.js

```diff
@@ -1,97 +1,132 @@
+    cancelCallback: null,
+    error: null,
+  });
 
+  const queryResultInExecution = useRef(null);
+  // Clear executing queryResult when component is unmounted to avoid errors
   useEffect(() => {
-    if (!isQueryExecuting) {
-      showNotificationMessageRef.current();
-    }
-  }, [isQueryExecuting]);
+    return () => {
+      queryResultInExecution.current = null;
+    };
+  }, []);
 
-  const executeQuery = useCallback(() => {
-    recordEvent("execute", "query", query.id);
-    setQueryResult(query.getQueryResult(0));
-    notifications.getPermissions();
-  }, [query]);
+  const executeQuery = useCallback(
+    (maxAge = 0, queryExecutor) => {
+      let newQueryResult;
+      if (queryExecutor) {
+        newQueryResult = queryExecutor();
+      } else {
+        newQueryResult = query.getQueryResult(maxAge);
+      }
 
-  const executeAdhocQuery = useCallback(
-    selectedQueryText => {
       recordEvent("execute", "query", query.id);
-      setQueryResult(query.getQueryResultByText(0, selectedQueryText));
       notifications.getPermissions();
+
+      queryResultInExecution.current = newQueryResult;
+
+      setExecutionState({
+        updatedAt: newQueryResult.getUpdatedAt(),
+        executionStatus: newQueryResult.getStatus(),
+        isExecuting: true,
+        cancelCallback: () => {
+          recordEvent("cancel_execute", "query", query.id);
+          setExecutionState({ isCancelling: true });
+          newQueryResult.cancelExecution();
+        },
+      });
+
+      const onStatusChange = status => {
+        if (queryResultInExecution.current === newQueryResult) {
+          setExecutionState({ updatedAt: newQueryResult.getUpdatedAt(), executionStatus: status });
+        }
+      };
+
+      newQueryResult
+        .toPromise(onStatusChange)
+        .then(queryResult => {
+          if (queryResultInExecution.current === newQueryResult) {
+            // TODO: this should probably belong in the QueryEditor page.
+            if (queryResult && queryResult.query_result.query === query.query) {
+              query.latest_query_data_id = queryResult.getId();
+              query.queryResult = queryResult;
+            }
+
+            if (executionState.loadedInitialResults) {
+              notifications.showNotification("Redash", `${query.name} updated.`);
+            }
+
+            setExecutionState({
+              queryResult,
+              loadedInitialResults: true,
+              error: null,
+              isExecuting: false,
+              isCancelling: false,
+              executionStatus: null,
+            });
+          }
+        })
+        .catch(queryResult => {
+          if (queryResultInExecution.current === newQueryResult) {
+            if (executionState.loadedInitialResults) {
+              notifications.showNotification("Redash", `${query.name} failed to run: ${queryResult.getError()}`);
+            }
+
+            setExecutionState({
+              queryResult,
```

Comment:
There's also the option of keeping old queryResults when a query has an error, but you were thinking about an Error State?

Reply:
🤔 not sure what I had in mind here. Will experiment with both options and decide.